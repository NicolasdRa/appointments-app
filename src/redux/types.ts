export interface User {
  id: string;
  firstName: string;
  lastName: string;
  email: string;
  memberSince: string;
  dateOfBirth: string;
  insurance: string;
  img: string;
}

export interface Doctor {
  id: string;
  title?: string;
  firstName?: string;
  lastName?: string;
  gender?: string;
  email?: string;
  dateOfBirth?: string;
  registerDate?: string;
  phone?: string;
  picture?: string;
  location?: string;
}

export interface Appointment {
  id: string;
  title: string;
  date: string;
  time: string;
  user: string;
  doctor?: string;
  description?: string;
}
