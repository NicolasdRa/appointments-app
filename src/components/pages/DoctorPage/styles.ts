import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';
import DummyMap from '../../../assets/images/dummyMap.jpg';

export const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    loader: {
      display: 'flex',
      margin: '35vh auto',
    },

    container: {
      display: 'flex',
      flexDirection: 'column',
      flex: '0 0 auto',
      width: '100%',
    },

    header: {
      backgroundColor: theme.palette.secondary.main,
      height: '18vh',
      borderRadius: '6px',
      boxShadow:
        '0 5px 15px -8px rgb(0 0 0 / 24%), 0 8px 10px -5px rgb(0 0 0 / 20%)',
    },

    content: {
      display: 'flex',
      flexDirection: 'column',
      margin: '0 1rem',
    },

    userImg: {
      boxShadow:
        '0 5px 15px -8px rgb(0 0 0 / 24%), 0 8px 10px -5px rgb(0 0 0 / 20%)',
      borderRadius: '6px',
      height: 'auto',
      maxWidth: '100%',
      margin: '2rem 2rem 0 1rem',
    },

    user_info: {
      display: 'flex',
      flexDirection: 'column',
      justifyContent: 'center',
      margin: '3rem 1rem',
    },

    userFullName: {
      color: theme.palette.common.chalk,
    },

    discipline: {},

    infoBlock: {
      padding: '1rem',
      minHeight: '57vh',
    },

    gallery: {
      display: 'flex',
      flexDirection: 'column',
      marginTop: '3rem',
      flex: '0 0 auto',
    },

    text: {
      display: 'flex',
      flexDirection: 'column',
      flex: '0 0 auto',
      lineHeight: 1.8,
      margin: '.5rem',
    },

    info: {
      display: 'flex',
      flexDirection: 'row',
      justifyContent: 'space-between',
      alignItems: 'center',
      marginTop: '-3rem',
      padding: '0 1rem 0 0',
    },

    bookBtn: {
      display: 'flex',
      alignItems: 'center',
      width: '14rem',
      margin: '1.5rem auto',
    },

    divider: {
      margin: '1rem 0 .5rem 0',
    },

    mapContainer: {
      width: '100%',
      height: '25vh',
    },

    map: {
      height: '100%',
      width: 'auto',
      padding: '.5rem',
      borderRadius: '12px',
    },
  }),
);
