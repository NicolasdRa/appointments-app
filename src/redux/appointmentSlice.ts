import { createSlice, PayloadAction } from '@reduxjs/toolkit';

import type { RootState } from './store';
import { Appointment } from './types';
import { initialAppointments } from '../assets/data';

// SLICE

// type for the slice state
interface AppointmentsState {
  data: Appointment[];
}

// initial state
const initialState: AppointmentsState = { data: initialAppointments };

export const appointmentSlice = createSlice({
  name: 'appointments',
  initialState,
  reducers: {
    create: (state, action: PayloadAction<Appointment>) => {
      state.data.push(action.payload);
    },
    cancel: (state, action: PayloadAction<Appointment>) => {
      const filtered = state.data.filter(
        (item) => item.id !== action.payload.id,
      );
      state.data = filtered;
    },
  },
});

// ACTIONS
export const { create, cancel } = appointmentSlice.actions;

// SELECTORS
export const selectAppointments = (state: RootState) => state.appointments.data;

// TODO: create selector that returns appointments by user

export default appointmentSlice.reducer;
