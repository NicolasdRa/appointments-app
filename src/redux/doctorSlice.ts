import { createSlice, createAsyncThunk, PayloadAction } from '@reduxjs/toolkit';
import axios from 'axios';

import type { RootState } from './store';
import { Doctor } from './types';
import { APP_ID, baseURL } from '../constants';

// THUNKS
export const fetchDoctors = createAsyncThunk('doctors/fetch', async () => {
  const res = await axios({
    method: 'GET',
    url: `${baseURL}`,
    headers: {
      'app-id': `${APP_ID}`,
    },
    responseType: 'json',
  });

  return res.data.data as Doctor[];
});

// SLICE
export interface DoctorState {
  loading: 'idle' | 'pending' | 'succeeded' | 'failed';
  data: Doctor[] | [];
  current: Doctor | undefined;
}

const initialState: DoctorState = {
  loading: 'idle',
  data: [],
  current: undefined,
};

const doctorSlice = createSlice({
  name: 'doctors',
  initialState,
  reducers: {
    setCurrent: (state, action: PayloadAction<Doctor>) => {
      state.current = action.payload;
    },
  },
  extraReducers: (builder) => {
    builder.addCase(fetchDoctors.pending, (state, action) => {
      state.loading = 'pending';
    });
    builder.addCase(fetchDoctors.fulfilled, (state, action) => {
      state.loading = 'succeeded';
      state.data = action.payload;
    });
    builder.addCase(fetchDoctors.rejected, (state, action) => {
      state.loading = 'failed';
    });
  },
});

// ACTIONS
export const { setCurrent } = doctorSlice.actions;

// SELECTORS
export const selectDoctors = (state: RootState) => state.doctors.data;

export const selectCurrentDoctor = (state: RootState) => state.doctors.current;

export default doctorSlice.reducer;
