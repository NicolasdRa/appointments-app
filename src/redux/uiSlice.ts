import { createSlice, PayloadAction } from '@reduxjs/toolkit';

import type { RootState } from './store';

// SLICE
interface UiState {
  formOpen: boolean;
}

const initialState: UiState = {
  formOpen: false,
};

const uiSlice = createSlice({
  name: 'ui',
  initialState,
  reducers: {
    toogleFormOpen: (state, action: PayloadAction) => {
      const open = state.formOpen;
      state.formOpen = !open;
    },
  },
});

// ACTIONS
export const { toogleFormOpen } = uiSlice.actions;

// SELECTORS
export const selectFormOpen = (state: RootState) => state.ui.formOpen;

export default uiSlice.reducer;
