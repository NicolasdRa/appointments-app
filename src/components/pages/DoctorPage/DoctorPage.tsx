import { useParams, Redirect } from 'react-router-dom';

import {
	Box,
	Button,
	Divider,
	Grid,
	Hidden,
	Typography,
} from '@material-ui/core';

import { Header } from '../../ui/Header/Header';
import { Footer } from '../../ui/Footer/Footer';
import { BottomNav } from '../../ui/BottomNav/BottomNav';

import { useStyles } from './styles';

import { selectDoctors, setCurrent } from '../../../redux/doctorSlice';
import { useAppDispatch, useAppSelector } from '../../../hooks/hooks';
import { Doctor } from '../../../redux/types';
import DummyMap from '../../../assets/images/dummyMap.jpg';
import { toogleFormOpen } from '../../../redux/uiSlice';

interface NameParams {
	name: string;
}

export const DoctorPage = () => {
	const classes = useStyles();
	const dispatch = useAppDispatch();

	const { name } = useParams<NameParams>();

	const doctors: Doctor[] = useAppSelector(selectDoctors);

	const doctor: Doctor = doctors.filter(
		(doctor) => doctor.name.last === name,
	)[0];

	if (!doctor) {
		return <Redirect to='/' />;
	}

	const {
		name: { first, last },
		picture: { medium },
	} = doctor;

	const handleOpenForm = (): void => {
		dispatch(setCurrent(doctor));
		dispatch(toogleFormOpen());
	};

	return (
		<Box>
			<Header />
			<Box className={classes.main}>
				<Box className={classes.imageContainer}>
					<Grid
						container
						direction='row'
						justify='flex-start'
						alignItems='flex-start'>
						<Grid item xs={6} sm={2}>
							<img
								src={medium}
								alt={first + '' + last}
								className={classes.userImg}
							/>
						</Grid>
						<Grid item xs={6} lg={6}>
							<Box className={classes.user_info}>
								<Typography
									variant='h5'
									className={classes.userFullName}>
									Dr. {first} {last}
								</Typography>
								<Typography
									variant='body2'
									color='textSecondary'
									component='p'
									className={classes.discipline}>
									general practitioner
								</Typography>
							</Box>
						</Grid>
					</Grid>
				</Box>
				<Box className={classes.content}>
					<Box className={classes.infoBlock}>
						<Typography
							variant='subtitle1'
							className={classes.text}>
							DOCTOR PROFILE
						</Typography>
						<Typography variant='body2' className={classes.text}>
							Dear patients, welcome to my general practice in
							Berlin! Here I am happy to advise you on all acute
							and chronic health complaints with regard to the
							appropriate therapy. Please make an appointment
							directly online. My team and I welcome you to our
							friendly practice. Here I am your first point of
							contact for all general health concerns and
							complaints and I am also at your side in the context
							of follow-up examinations and follow-up treatments
							after any illnesses. Your Dr. {last}
						</Typography>
						<Divider className={classes.divider} />
						<Grid container>
							<Grid item xs={12} lg={6}>
								<Typography
									variant='subtitle1'
									className={classes.subtitle}>
									Languages
								</Typography>
								<Typography
									variant='body2'
									className={classes.text}>
									English, German and Spanish
								</Typography>
							</Grid>
							<Grid item xs={12} lg={6}>
								<Typography
									variant='subtitle1'
									className={classes.subtitle}>
									Website
								</Typography>
								<Typography
									variant='body2'
									className={classes.text}>
									wwww.dr{first}
									{last}.com
								</Typography>
							</Grid>
							<Grid item xs={12} lg={12}>
								<Divider className={classes.divider} />
								<Button
									variant='contained'
									className={classes.bookBtn}
									onClick={handleOpenForm}>
									BOOK AN APPOINTMENT
								</Button>
							</Grid>
						</Grid>
					</Box>
					<Box className={classes.infoBlock}>
						<Typography
							variant='subtitle1'
							className={classes.text}>
							OPENING HOURS & CONTACT{' '}
						</Typography>
						<Grid container>
							<Grid item xs={12} lg={6}>
								<Typography
									variant='subtitle1'
									className={classes.subtitleUp}>
									Opening hours
								</Typography>
								<Typography
									variant='body2'
									className={classes.text}>
									Mo, Wed: 8:30am - 5:00pm
								</Typography>
								<Typography
									variant='body2'
									className={classes.text}>
									Tue, Thu: 8:30am - 1:00pm
								</Typography>
								<Typography
									variant='body2'
									className={classes.text}>
									Fri: 8:30am - 3pm
								</Typography>
								<Typography
									variant='body2'
									className={classes.text}>
									Sat: 8:30am - 1pm
								</Typography>
							</Grid>
							<Grid item xs={12} lg={6}>
								<Typography
									variant='subtitle1'
									className={classes.subtitleUp}>
									Phone number
								</Typography>
								<Typography
									variant='body2'
									className={classes.text}>
									030 89405090
								</Typography>
								<Typography
									variant='subtitle1'
									className={classes.subtitle}>
									Emergency contact
								</Typography>
								<Typography
									variant='body2'
									className={classes.text}>
									In emergencies, please contact 030 116117
									(emergency medical service)
								</Typography>
							</Grid>
						</Grid>

						<Grid container>
							<Grid item xs={12} lg={6}>
								<Typography
									variant='subtitle1'
									className={classes.subtitle}>
									Location
								</Typography>
								<Box className={classes.mapContainer}>
									<img
										src={DummyMap}
										alt='location'
										className={classes.map}
									/>
								</Box>
							</Grid>

							<Grid item xs={12} lg={6}>
								<Typography
									variant='subtitle1'
									className={classes.subtitle}>
									Access information
								</Typography>
								<Typography
									variant='body2'
									className={classes.text}>
									{first} {last}'s practice
								</Typography>
								<Typography
									variant='body2'
									className={classes.text}>
									Torstrasse 222, 11111 Berlin
								</Typography>

								<Typography
									variant='body2'
									className={classes.text}>
									U-Bahn - U Sch√∂nleinstr. (Line U8)
								</Typography>
								<Typography
									variant='body2'
									className={classes.text}>
									Ground floor barrier-free
								</Typography>
							</Grid>
						</Grid>
					</Box>
				</Box>
			</Box>
			<Footer />
			<Hidden mdUp>
				<BottomNav />
			</Hidden>
		</Box>
	);
};
