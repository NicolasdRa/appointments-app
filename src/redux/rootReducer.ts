import { combineReducers } from 'redux';

import doctorSlice from './doctorSlice';
import appointmentSlice from './appointmentSlice';
import uiSlice from './uiSlice';

const rootReducer = combineReducers({
  doctors: doctorSlice,
  appointments: appointmentSlice,
  ui: uiSlice,
});

export default rootReducer;
