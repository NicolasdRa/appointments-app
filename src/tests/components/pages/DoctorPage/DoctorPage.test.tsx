import { DoctorPage } from '../../../../components/pages/DoctorPage/DoctorPage';
import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import { store } from '../../../../redux/store';

describe('tests on <DoctorPage/>', () => {
  test('should render <DoctorPage/> correctly', () => {
    const wrapper = shallow(
      <Provider store={store}>
        <DoctorPage />
      </Provider>,
    );

    expect(wrapper).toMatchSnapshot();
  });
});
