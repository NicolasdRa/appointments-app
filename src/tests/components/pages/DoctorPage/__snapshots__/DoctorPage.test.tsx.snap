// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests on <DoctorPage/> should render <DoctorPage/> correctly 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "@@observable": [Function],
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "@@observable": [Function],
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <DoctorPage />
</ContextProvider>
`;
